clients/src/main/java/org/apache/kafka/common/network/Authenticator.java

```diff
@@ -54,4 +55,104 @@ default void handleAuthenticationFailure() throws IOException {
+     *             security configuration errors
+     * @throws IOException
+     *             if read/write fails due to an I/O error
+     */
+    default void reauthenticate(ReauthenticationContext reauthenticationContext) throws IOException {
+        // empty
+    }
+
+    /**
+     * Return the time, in milliseconds since the epoch, when the session began.
+     * Valid on both client- and server-side.
+     * 
+     * @return the time, in milliseconds since the epoch, when the session began
+     * @throws IllegalStateException
+     *             if the session is not established
+     * @see #complete()
+     */
+    long sessionBeginTimeMs();
+
+    /**
+     * Return the client-side session expiration time, in milliseconds since the
+     * epoch, after which re-authentication must occur before the connection is used
+     * for anything else; or return null if the session has no expiration or it is
+     * on the server-side. Note that any non-null value does not take into account
+     * either the possibility of clock drift between the client and the server or
+     * the inherent latency introduced by the fact that messages have to traverse
+     * the network between the client and the server. It is therefore prudent to
+     * assume the session expires at some point before the given time (e.g. 90% of
+     * the way between the {@link #sessionBeginTimeMs()} and this time).
+     * 
+     * @return the client-side session expiration time, in milliseconds since the
+     *         epoch, after which re-authentication must occur before the connection
+     *         is used for anything else; or return null if the session has no
+     *         expiration
+     */
+    default Long sessionExpirationTimeMs() {
+        return null;
+    }
+
+    /**
+     * Return the client-side {@link NetworkReceive} responses that arrived during
+     * re-authentication that are unrelated to re-authentication, if any, otherwise
+     * null. These correspond to requests sent prior to the beginning of
+     * re-authentication; the requests were made when the channel was successfully
+     * authenticated, and the responses arrived during the re-authentication
+     * process.
+     * 
+     * @return the client-side {@link NetworkReceive} responses that arrived during
+     *         re-authentication that are unrelated to re-authentication, if any,
+     *         otherwise null
+     */
+    default Deque<NetworkReceive> getAndClearResponsesReceivedDuringReauthentication() {
+        return null;
+    }
+    
+    /**
+     * Return true if this is a server-side authenticator and the connected client
+     * supports re-authentication, otherwise false
+     * 
+     * @return true if this is a server-side authenticator and the connected client
+     *         supports re-authentication, otherwise false
+     */
+    default boolean clientSupportsReauthentication() {
```

Comment:
👍 I've renamed this to be `boolean connectedClientSupportsReauthentication()` which is clear and no longer is close to any other method names since I eliminated the other two that were close as described above.